pkg/controller/controller.go

```diff
@@ -0,0 +1,288 @@
+
+	for _, remoteClusters := range c.remoteClusters {
+		remoteClusters.close()
+	}
+}
+
+func (c *LightHouseController) OnAdd(clusterID string, kubeConfig *rest.Config) {
+	klog.Infof("adding cluster: %s", clusterID)
+
+	clientSet, err := c.newClientset(kubeConfig)
+	if err != nil {
+		klog.Errorf("error creating clientset for cluster %s: %s", clusterID, err.Error())
+		return
+	}
+
+	//startNewServiceWatcher acquires the lock before accessing the map.
+	c.startNewServiceWatcher(clusterID, clientSet)
+}
+
+func (c *LightHouseController) OnUpdate(clusterID string, kubeConfig *rest.Config) {
+	//TODO asuryana need to implement
+	klog.Infof("updating cluster: %s", clusterID)
+	klog.Fatalf("Not implemented yet")
+}
+
+func (c *LightHouseController) OnRemove(clusterID string) {
+	c.processingMutex.Lock()
+	defer c.processingMutex.Unlock()
+	c.removeServiceWatcher(clusterID)
+}
+
+func (c *LightHouseController) startNewServiceWatcher(clusterID string, clientSet kubernetes.Interface) {
+	serviceInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options meta_v1.ListOptions) (runtime.Object, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options meta_v1.ListOptions) (watch.Interface, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Service{},
+		0,
+		cache.Indexers{},
+	)
+	remoteCluster := &remoteCluster{
+		stopCh:           make(chan struct{}),
+		clusterID:        clusterID,
+		serviceInformer:  serviceInformer,
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		remoteServiceMap: c.remoteServiceMap,
+		federator:        c.federator,
+		processingMutex:  c.processingMutex,
+	}
+	c.processingMutex.Lock()
+	defer c.processingMutex.Unlock()
+	c.remoteClusters[clusterID] = remoteCluster
+
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go remoteCluster.run()
+}
+
+func (c *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := c.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(c.remoteClusters, clusterID)
+	}
+}
+
+func (rc *remoteCluster) close() {
+	close(rc.stopCh)
+	rc.queue.ShutDown()
+	klog.Infof(" watcher queue shutdown for cluster: %s", rc.clusterID)
+}
+
+func (c *remoteCluster) run() {
+	defer utilruntime.HandleCrash()
+
+	go c.serviceInformer.Run(c.stopCh)
+
+	if !cache.WaitForCacheSync(c.stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
```

Comment:
I don't think you should return here as it will elide the call to `c.runWorker()` below with no way to recover. Most likely I think this could fail if the remote API server isn't available in which case I believe it could connect later and start the watch. That said, I don't think you really need to call `WaitForCacheSync` at all for your use case.

Reply:
Removed.